internal/app/export/service.go

```diff
@@ -6,80 +6,59 @@ package export
 import (
 	"fmt"
 
-	"github.com/gogo/protobuf/proto"
-	"github.com/mainflux/mainflux"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/mainflux/export/internal/pkg/routes"
+	"github.com/mainflux/export/internal/pkg/routes/mfx"
+	"github.com/mainflux/export/pkg/config"
 	log "github.com/mainflux/mainflux/logger"
-	"github.com/mainflux/mainflux/transformers"
 	nats "github.com/nats-io/nats.go"
 )
 
 type Service interface {
-	Start(string) error
-	Consume(m *nats.Msg)
-	Export(msgs ...interface{})
+	Start(string)
 }
 
 var _ Service = (*exporter)(nil)
 
 type exporter struct {
-	Nc          *nats.Conn
-	Channels    map[string]bool
-	Repo        MessageRepository
-	Transformer transformers.Transformer
-	Logger      log.Logger
-	Cons        func(*nats.Msg)
+	NatsConn *nats.Conn
+	Mqtt     mqtt.Client
+	Logger   log.Logger
+	Cfg      config.Config
+	Routes   []routes.RouteIF
 }
 
 // New create new instance of export service
-func New(nc *nats.Conn, repo MessageRepository, transformer transformers.Transformer, channels map[string]bool, fConsume func(*nats.Msg), logger log.Logger) Service {
-
-	c := exporter{
-		Nc:          nc,
-		Channels:    channels,
-		Repo:        repo,
-		Transformer: transformer,
-		Logger:      logger,
-		Cons:        fConsume,
+func New(nc *nats.Conn, mqtt mqtt.Client, c config.Config, logger log.Logger) Service {
+	routes := make([]routes.RouteIF, 0)
+	e := exporter{
+		NatsConn: nc,
+		Mqtt:     mqtt,
+		Logger:   logger,
+		Cfg:      c,
+		Routes:   routes,
 	}
 
-	if fConsume == nil {
-		c.Cons = c.Consume
-	}
-	return &c
+	return &e
 }
 
-// Start method starts consuming messages received from NATS.
-// This method transforms messages to SenML format before
-// using MessageRepository to store them.
-func (c *exporter) Start(queue string) error {
-	_, err := c.Nc.QueueSubscribe(mainflux.InputChannels, queue, c.Cons)
-	return err
-}
-
-func (c *exporter) Consume(m *nats.Msg) {
-	msg := mainflux.Message{}
-	err := proto.Unmarshal(m.Data, &msg)
-	if err == nil {
-		msgs := []interface{}{}
-		msgs = append(msgs, msg)
-		c.Export(msgs...)
-		return
-	}
-	c.Logger.Warn(fmt.Sprintf("Failed to unmarshal received message: %s", err))
-	return
-}
-
-func (c *exporter) Export(msgs ...interface{}) {
-	if err := c.Repo.Publish(msgs...); err != nil {
-		c.Logger.Warn(fmt.Sprintf("Failed to save message: %s", err))
-		return
-	}
-}
+// Start method starts consuming messages received from NATS
+// and makes routes according to the configuration file.
+// Routes export messages to mqtt.
+func (e *exporter) Start(queue string) {
+	var route routes.RouteIF
+	for _, r := range e.Cfg.Routes {
+		switch r.Type {
+		case "mfx":
+			route = mfx.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
```

Comment:
Think about creating `Subscriber` interface that will be implemented using `nats` package, and `Publisher` interface that will be implemented using `mqtt` package. Then, you can use these interfaces from the `service` in order to read and forward messages. Also, maybe you can separate `MQTT` client and logger from `Route` in this way.

Reply:
https://github.com/mainflux/export/issues/8